src/uct/tcp/tcp_ep.c

```diff
@@ -616,16 +620,21 @@ static unsigned uct_tcp_ep_progress_data_tx(uct_tcp_ep_t *ep)
 {
     unsigned count = 0;
     size_t sent_length;
+    int ret;
 
     ucs_trace_func("ep=%p", ep);
 
     if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
-        if (!(ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX))) {
-            count += uct_tcp_ep_send(ep, &sent_length);
-        } else {
-            count += uct_tcp_ep_sendv(ep, &sent_length);
+        ret = (!(ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX)) ?
```

Comment:
There is a kind of mess of return values:
`count` should be changed in any case if something handled including `UCS_OK` and unrecoverable errors. 
`count` shouldn't be changed if return value is smth like `UCS_NO_RESOURCES` , `UCS_NO_PROGRESS`, `EAGAIN` etc.
I would suggest to use one from existing return value rules:
1) `SEND(2)` (then need to return `ssize_t`, check against `-1` (+ `errno` in case of failure) and no need `length` parameter)
2) or `ucs_status_t` (might require more checks)

Reply:
Implemented 1st variant, but return UCS status instead of `-1`, because a negative value is an error. So, we don't need to check `errno`

Reply:
then you can't use `UCS_STATUS_IS_ERR` and other checks against `UCS_*` status values

Reply:
> then you can't use `UCS_STATUS_IS_ERR` and other checks against `UCS_*` status values

why?

Reply:
because it mixes 2 different approaches

Reply:
> because it mixes 2 different approaches

checking for `< 0` now:
- `< 0` - error occured
- `>= 0` - normal case, where `> 0` means some data were sent